source/Halibut/Transport/Proxy/HttpProxyClient.cs

```diff
@@ -347,25 +347,31 @@ async Task SendConnectionCommandAsync(string host, int port, CancellationToken c
             // send the connect request
             await stream.WriteAsync(request, 0, request.Length, cancellationToken);
 
-            // wait for the proxy server to respond
-            await WaitForDataAsync(stream, cancellationToken);
-
             // PROXY SERVER RESPONSE
             // =======================================================================
             //HTTP/1.0 200 Connection Established<CR><LF>
             //[.... other HTTP header lines ending with <CR><LF>..
             //ignore all of them]
             //<CR><LF>    // Last Empty Line
-
-            // create an byte response array  
-            var response = new byte[TcpClient.ReceiveBufferSize];
+  
+            
             var sbuilder = new StringBuilder();
 
-            do
+            var existingTimeout = stream.ReadTimeout;
+            stream.ReadTimeout = WAIT_FOR_DATA_TIMEOUT;
```

Comment:
Proxy doesn't work with websockets so it can require that the stream supports timeouts. Which is what we should always have been doing.

Reply:
Have you confirmed it doesn't work, I never managed to get it to work but does something magic happen in Server and it does just work?

Reply:
Will this somehow cause Server/Tentacle to blow up if a proxy is configured for WebSockets?

Reply:
Oh I must have understood not getting it to work to mean it doesn't work in halibut.

Looking at the code it looks like it has support for it:

```
WebSocketConnectionFactory.cs:
async Task<ClientWebSocket> CreateConnectedClientAsync(ServiceEndPoint serviceEndpoint, CancellationToken cancellationToken)
        {
            if (!serviceEndpoint.IsWebSocketEndpoint)
                throw new Exception("Only wss:// endpoints are supported");
            var client = new ClientWebSocket();
            ...
            if (serviceEndpoint.Proxy != null)
                client.Options.Proxy = new WebSocketProxy(serviceEndpoint.Proxy);
```

Note that the proxy options are passed directly to the ClientWebSocket. It doesn't look like [HttpProxyClient.cs](https://github.com/OctopusDeploy/Halibut/pull/468/files#diff-f236e7eb8c2b940a8385632e889fc9145bed41a34b6d6f46be77e4858773be37) is involved at all.

![image](https://github.com/OctopusDeploy/Halibut/assets/7076477/6ffbd846-0d9e-4939-91da-30cb67f68682)

Reply:
I also am finding that http proxies do not work with websockets. Interestingly if I put break points into the methods of `WebSocketProxy` a halibut class which implements a dotnet interface `IWebProxy` those break points are never hit!

The `HttpProxyClient` is not involved in websockets as the code expects `ClientWebSocket` a dotnet class to do the work of connecting via the proxy (which it doesn't appear to do).

Reply:
That's the same behaviour as I observed. It does seem broken on Microsoft's side!!

Reply:
Should we safeguard anyway and not assume the stream can timeout? E.g. it gets fixed and suddenly we have a bug

Reply:
We could wrap it at this point with a SupportsTimeoutStream and then a NetworkTimeoutStream if it doesn't support timeouts as an interim fix while the WebSopcketStream doesn't support timeouts

Reply:
I think it should be fine since we always have a network stream:
![image](https://github.com/OctopusDeploy/Halibut/assets/7076477/feb73faa-68ab-49fb-9c59-aae0ece628f0)